### ISSUE TITLE
Complete the Bullet C-API (was googlecode  Issue 43)

### ISSUE THREAD

**Issue by erwincoumans** [bulletphysics/bullet3#130]
See https://code.google.com/p/bullet/issues/detail?id=43&colspec=Modified%20ID%20Type%20Stars%20Status%20Owner%20Summary

**Comment by chino** [comment]
I really hope bullet 3 api will have a supported c api.

Lack of a stable supported fully featured wrapper has been biggest slow down with using Bullet in my non c++ projects.

Is the bullet 3 api going to be able to downgrade back to the cpu if the video card card isn't capable enough?

Thanks for all the great work.

**Comment by AndresTraks** [comment]
I've also been working on a C API as part of a P/Invoke .NET wrapper. It's pretty much complete for bullet 2 and I've planned to start with bullet 3 this summer. https://github.com/AndresTraks/BulletSharpPInvoke/tree/master/libbulletc

It's a bit different though as it's first and foremost designed to be an intermediate layer between bullet and other higher level wrappers instead of being a natural C API. The code is completely independent from bullet, so nothing needs to be modified in bullet itself.

Here's the idea of how it currently works:
```cpp
//btCollisionShape_wrap.h
#ifndef BT_COLLISION_SHAPE_H
#define btCollisionShape void
#endif
EXPORT void btCollisionShape_delete(btCollisionShape* obj);
```
```cpp
//btSphereShape_wrap.h
#ifndef BT_SPHERE_MINKOWSKI_H
#define btSphereShape void
#endif
EXPORT btSphereShape* btSphereShape_new(btScalar radius);
EXPORT btScalar btSphereShape_getRadius(btSphereShape* obj);
```
```cpp
//btCollisionShape_wrap.cpp
#include <BulletCollision/CollisionShapes/btCollisionShape.h>
#include "btCollisionShape_wrap.h"

void btCollisionShape_delete(btCollisionShape* obj)
{
	delete obj;
}
```
```cpp
//btSphereShape_wrap.cpp
#include <BulletCollision/CollisionShapes/btSphereShape.h>
#include "btSphereShape_wrap.h"

btSphereShape* btSphereShape_new(btScalar radius)
{
	return new btSphereShape(radius);
}

btScalar btSphereShape_getRadius(btSphereShape* obj)
{
	return obj->getRadius();
}
```
```cpp
//bulletc.h
#include "btCollisionShape_wrap.h"
#include "btSphereShape_wrap.h"
```
```cpp
//test.cpp
#include <bulletc.h>

btSphereShape* shape = btSphereShape_new(1.0f);
btScalar radius = btSphereShape_getRadius(shape);
btCollisionShape_delete(shape);
```

**Comment by Tal500** [comment]
AndresTraks,
As you might see, I aim to have C API embedded in Bullet 3.
For doing this, the big challenge is to make LinearMath library of Bullet compatible with C, with no loose of optimizations. (Actually, I theoretically improved optimization since I use a generic "restricted" pointer specifier when possible

After this, the C API might be easily written in full Bullet 3 API.

Do you think we should combine our efforts?

**Comment by AndresTraks** [comment]
Yes, I can surely help. My interest is that Bullet could be compiled into a .dll/.so library with the C symbols exported. This is a standard way of writing wrappers for other languages like C# and Lua.

**Comment by ebkalderon** [comment]
Bump. Any progress on this since the last update? I am very interested in writing up-to-date Bullet bindings for Rust, which would require a C FFI.

**Comment by Tal500** [comment]
Sadly I'm very busy these days, but I didn't forget it, just need some time and help.

As said above, I focused on first supporting LinearMath before the rest of Bullet.
I have completed the API for Bullet vectors, working on the rest.

You might use this, and extend this binding to the Bullet API function in your project, and send your implementation for updating the C API.

I would like however some help on finishing LinearMath implementation, contact me if you're interested.

**Comment by Tal500** [comment]
Hi all,

I've completed basic Bullet 2 LinearMath C API.

Now I want to write _embedded_ C API for Bullet 2 Core(collision&dynamics). By _embedded_ I mean that the function deceleration are in the original header, and its implementation is in the original source.

(AndresTraks especially, I want you to answer this :)
How do you recommend me to merge AndresTraks work there? Where to start?
Does any one want to join?

**Comment by bmcorser** [comment]
@ebkalderon :point_up:

**Comment by Tal500** [comment]
After some serious thinking, I decided to pause my effort on the Bullet fork with embedded support, until some merging or decision from official Bullet.

As I described many before, I finished making LinearMath compatible with C. I thought it was crucial if we want C API inside the official Bullet API. If ODE did only C API, Bullet can do both C and C++ API.

Now, after this LinearMath C API infrastructure, writing a C API in the same headers should be not difficult. I'm not going to do so if it won't be merged, or a strategic decision will be made.

Even known it's only for Bullet 2 right now I still think it's important, since it's the widely used version right now.

Wait for your response.

**Comment by CapsAdmin** [comment]
@Tal500 Maybe you should make a pull request and propose a first step towards a C api.

**Comment by Tal500** [comment]
@CapsAdmin, as you suggested I created this pull request: https://github.com/bulletphysics/bullet3/pull/636

**Comment by rweichler** [comment]
@kyle-emmerich @erwincoumans just FYI, im working on a C wrapper, with the end goal to be used with the LuaJIT ffi:

https://github.com/rweichler/bullet_luajit/blob/master/bullet_wrapper.cpp

**Comment by erwincoumans** [comment]
Thanks for sharing the various projects and thoughts about a Bullet 2.x specific C-API. We have pybullet python bindings for bullet3/examples/SharedMemory/PhysicsClientC_API.h and it would be straight-forward to add Lua bindings the that C-API as well. The move to Bullet 3.x and 4.x is going to happen, and don't want to extend the life of the 2.x API, even if that is the current most active version. Let's keep this issue open for discussion, some people may like to use the Bullet 2.x specific C-APIs.

**Comment by EvilPudding** [comment]
How is this going along? Is there any viable C API nowadays?

**Comment by Tal500** [comment]
EvilPudding, I stopped the work since Erwin didn't accept it. He said(as much as I understand) that the new C API will be part of Bullet3 API.

**Comment by EvilPudding** [comment]
> that the new C API will be part of Bullet3 API

What does that mean? And is it available right now?

**Comment by erwincoumans** [comment]
Yes, the new C API is here:
https://github.com/bulletphysics/bullet3/blob/master/examples/SharedMemory/PhysicsClientC_API.h

It is used in PyBullet, see the [quickstart guide here](https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#heading=h.2ye70wns7io3).
You can also use the[ C++ API ](https://github.com/erwincoumans/bullet3/blob/master/examples/SharedMemory/b3RobotSimulatorClientAPI_NoGUI.h)on top of this C-API. As you can see, it is more high-level.
This is also available as 'BulletRobotics' using cmake (in Extras).

**Comment by EvilPudding** [comment]
@erwincoumans Thanks a lot!

**Comment by Tal500** [comment]
Hi Erwin
It seems like this C API has two major disadvantages:
* Is it following Singelton design pattern? I.e. only one global world.
* User pass handlers, but no user_data approach.
I mean, you can have a callback foo(args...,void* user_data), so handlers can get data locally, not by global memory.

**Comment by EvilPudding** [comment]
@Tal500 The good thing about that it's backwards compatible, meaning you can pass a function callback whose signature does not receive the `user_data`, libs like gtk even encourage this.

**Comment by erwincoumans** [comment]
@Tal500 There is no global world: you can create multiple physics clients/servers and each server has its own world.
You can set user-data if desired, for each link. This API doesn't use callbacks: you create one or more physics clients/servers, and you create objects, step the simulation, query for transforms, all using unique ids for each client/server.

**Comment by Tal500** [comment]
@erwincoumans I didn't understand the API at first, but now it's look great!

**Comment by EvilPudding** [comment]
@erwincoumans Why is the C API you linked in an example and is not installed by default? Do we have to copy the example API for each project and update it manually every time there's an update?

**Comment by EvilPudding** [comment]
I don't even know how to compile this... I tried to use the cmake file of the SharedMemory example, and I get the following error:

```
... bullet3/examples/SharedMemory/plugins/pdControlPlugin/pdControlPlugin.cpp:8:10: fatal error: LinearMath/btScalar.h: No such file or directory
 #include "LinearMath/btScalar.h"
```

Why does this file assume it is in the `src` folder? Am I building it wrong?

**Comment by EvilPudding** [comment]
Can anyone walk me through how to get a compiled library so I can use the C API? Why is this so mangled?

**Comment by erwincoumans** [comment]
@EvilPudding I recommend you go troll another open source project.

**Comment by erwincoumans** [comment]
@EvilPudding 
cmake should build the BulletRobotics library out-of-the-box, which includes the C and C++ API. It will be installed using the INSTALL_EXTRA_LIBS option.

App_RobotSimulator and App_HelloBulletRobotics are two example of using the C and C++ API. Both should build out-of-the box as well, using cmake.

**Comment by EvilPudding** [comment]
Thanks for the help, I'm sorry if this was obvious stuff, it's just it's a bit different from other libraries I've used before. Thanks for all the hard work.

**Comment by EvilPudding** [comment]
Ok, here's where I got it wrong, my package manager has an older version of the library which does not contain the C api... Sorry about that.

**Comment by EvilPudding** [comment]
@erwincoumans I understand if you don't have the time to answer, but I'm trying to compile a C file using the API, and I can't seem to link it with `libBulletRobotics.a`

If I just link it, `cc` outputs a bunch of `undefined reference to vtable ...`, and if I try to include it as a shared library using `-shared -fPIC` I get:

```
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../lib64/libBulletRobotics.a(LoadMeshFromCollada.o): warning: relocation against `_Z17readNodeHierarchyPN8tinyxml210XMLElementER9btHashMapI12btHashStringiER20btAlignedObjectArrayI23ColladaGraphicsInstanceERK11btMatrix4x4' in read-only section `.text'
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../lib64/libBulletRobotics.a(PhysicsServerCommandProcessor.o): relocation R_X86_64_PC32 against symbol `gVRTeleportPos1' can not be used when making a shared object; recompile with -fPIC
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: final link failed: Bad value
```

Again, I don't want to be a bother, so if you don't have the time, I understand if you don't answer.

**Comment by EvilPudding** [comment]
Turns out I needed to add `-DBUILD_SHARED_LIBS=ON` to the cmake options. Though the following symbols are still missing:

```
b3CreateInProcessPhysicsServerAndConnect
b3CreateInProcessPhysicsServerAndConnectMainThread
b3CreateInProcessPhysicsServerAndConnectSharedMemory
b3CreateInProcessPhysicsServerFromExistingExampleBrowserAndConnect3
```
I tried to search for them in the elf information in the installed libs and can't find where they ended up. If I don't use those functions it looks like it works fine.

This is a great library, and I'm very thankful for all the work put in here, thanks @erwincoumans
